---
title: "**Comparison of ONT-V1V9 vs. Illumina-V3V4 in PRJNA911189**"
output:
  html_document:
    highlight: 'haddock'
    # theme: paper
    df_print: paged
    # number_sections: true
    toc: true
    toc_float: true
    code_folding: hide
---

```{=html}
<style type="text/css">
.main-container {
  max-width: 100% !important;
  margin: auto;
}
</style>
```
```{r setup, include = FALSE}
knitr::opts_chunk$set(
    cache       = TRUE,     # if TRUE knitr will cache the results to reuse in future knits
    fig.width   = 11,       # the width for plots created by code chunk
    fig.height  = 8,       # the height for plots created by code chunk
    fig.align   = 'center', # how to align graphics in the final doc. 
    warning     = FALSE,
    message     = FALSE,
    comment     = NA
)
# knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
# getwd()
```

```{css, echo=FALSE, include = FALSE}
.header-section-number::after {
  content: ".";
}
```

# **Introduction**

This script compares two sequencing approaches for microbiome analysis of feces samples from cancer patients and healthy subjects through 16S rRNA. The approaches are full length V1V9 with Oxford Nanopore Technologies (ONT-V1V9) or partial length V3V4 with Illumina (Illumina-V3V4). Additionally, two databases (Silva and Default Emu) and three ONT dorado basecalling models (fast, hac, sup) are compared.

The main objective is to see if full length 16s rRNA improves the accuracy of previously described colorectal cancer biomarkers (Conde et. al 2023).

# **Libraries**

Load the neccessary libraries and set options such as a reproducible seed:

```{r message=FALSE, warning=FALSE,  class.source = 'fold-show'}
suppressMessages({
  # Libraries
  library(ggplot2)
  library(patchwork)
  # library(plotly)
  library(vegan)
  library(devtools)
  library(data.table)
  library(RColorBrewer)
  library(egg)
  library(randomcoloR)
  library(ggvenn)
  library(glue)
  library(tidyr)
  library(purrr)
  library(plyr)
  library(dplyr)
  library(phyloseq)
  library(ggalluvial)
  library(stringr)
  library(gtools)
  library(ggsignif)
  library(ggpubr)
  library(rstatix)
  library(ggmagnify)
  library(ggtext)
  library(ggrepel)
})

# Note: ANCOM-BC is loaded in its own section, this is because it loads tidytree and 
# creates an unavoidable warning/message each time a function in the the phyloseq package is used,
# which when outputting to the report makes it difficult to read

# General function
map_signif_level_func = function(x){
  # c("***"=0.001, "**"=0.01, "*"=0.05)
  if (x<=0.001){
    return("***")
  } else if (x<=0.01) {
    return("**")
  } else if (x<=0.05) {
    return("*")
  } else {
    return(NA)
  }
}

# Options
options(dplyr.summarise.inform = FALSE)
options(getClass.msg=FALSE)

# Set seed
set.seed(123)
```

Load a function to show pretty tables in HTML format:

```{r class.source = 'fold-show'}
create.table <- function(df, capt=NULL){
  DT::datatable(df, extensions = c('FixedColumns'),
                options = list(scrollX = TRUE, pageLength = 5), 
                caption=htmltools::tags$caption(
                  style = 'caption-side: bottom; text-align: center; color: black;',
                  htmltools::em(capt)
                ))
}

load_rdata <- function(fileName){
  # Loads an RData file, and returns it
  load(fileName)
  get(ls()[ls() != "fileName"])
}
```

# **Load data**

There are three parts that must be loaded for a phyloseq object, the features (feature), metadata and taxonomy information. The following sections will prepare the input and load them. The neccessary input for the script is specified here:

```{r class.source = 'fold-show'}
BASE_DIR = "/home/usuario/Proyectos/Results/ONT16S"
feature_table_file <- as.character(glue("/home/usuario/Proyectos/Results/ONT16S/pipeline/15_biom_conversion/feature_table.biom.json"))
metadata_file <- as.character(glue("/home/usuario/Proyectos/Results/ONT16S/pipeline/15_biom_conversion/metadata_table.tsv"))
taxmat_file <- as.character(glue("/home/usuario/Proyectos/Results/ONT16S/pipeline/15_biom_conversion/taxonomy_table.tsv"))
contamination_file <- as.character("/home/usuario/Proyectos/Results/ONT16S/data/contaminants.tsv")
# tree_file <- as.character(glue("{BASE_DIR}/output/qiime2/export_results/tree.nwk"))
output_folder <- as.character(glue("{BASE_DIR}/r_tests"))
biosample_col <- "biosample_name"
variables_of_interest <- c("platform", "database", "basecalling_model_name", "sample.type")
apply_controls <- FALSE
```

## Features

Load features and remove those which are empty (usually samples with bad quality or negative controls):

```{r}
# Transform biome data to JSON.biome
# biom convert -i in.biom -o out.biom.json --table-type="OTU table" --to-json

# Read
otu_df <- import_biom(feature_table_file)

# Remove columns/samples with all 0s (no data)
condition <- colSums(otu_df) != 0
otu_df_empty_cols <- colnames(otu_df)[!condition]
if (length(otu_df_empty_cols)>0){
  print(glue("WARNING: There are {length(otu_df_empty_cols)} empty columns in feature dataframe (no data, this can happen in controls) and they will be removed: '{paste(otu_df_empty_cols, collapse='; ')}'"))
}
otu_df <- otu_df[, condition]

create.table(head(otu_df), "Head of feature table.")
```

## Metadata

Metadata is read and initially processed here:

```{r}
# Read metadata
metadata <- read.delim(metadata_file, comment.char = "#")

# Check if there are columns not in otu_df besides the otu_df columns that were empty
condition <- metadata[,'sample.id'] %in% colnames(otu_df)
sample.ids.not.in.otus <- metadata[!condition,]$sample.id
if (length(setdiff(otu_df_empty_cols, sample.ids.not.in.otus))>0){
  warning("There are samples in metadata that do not exist in feature dataframe. This can happen if they were removed in the previous step.")
  print(setdiff(otu_df_empty_cols, sample.ids.not.in.otus))
}
metadata <- metadata[condition,] # eliminar muestras si no existen en otu_df
rownames(metadata) <- metadata[,'sample.id']

# Create column for grouping by variables_of_interest
if ("grouping_var" %in% names(metadata)){
  stop("Metadata already has grouping_var as column")
} else {
  metadata$grouping_var <- do.call(paste, c(metadata[variables_of_interest], sep=" - "))
}

# Make categorical columns
for (i in variables_of_interest){
  metadata[[i]] <- factor(metadata[[i]])
}

create.table(metadata, "Metadata table.")
```

## Taxonomy

The taxonomy is read and processed here, propagating the levels, filling NAs in taxa levels with the last known level:

```{r}
# Taxonomy must be split into columns
taxmat <- read.delim(taxmat_file, sep='\t')
rownames(taxmat) <- taxmat$feature_id
taxmat$feature_id <- NULL

# Propagate NAs
keep_unidentified <- function(tx_table){
  # Keep the ones that have one of c(NA, "", " ", "\t") in target_level column
  # and then find the last level that is known (can be Genus, Family...)
  # Assign to the target_level level the last known level + x__XXX NA
  # Family (last known)   Species
  # f__Lachnospiraceae    f__Lachnospiraceae NA
  elements_to_target <- colnames(tx_table)
  tx_table <- as.data.frame(tx_table)
  previous_level <- elements_to_target[1]
  for (level in elements_to_target[2:length(elements_to_target)]){
    empty_values <- tx_table[[level]] %in% c(NA, "", " ", "\t")
    tx_table[[level]][empty_values] <- paste(tx_table[[previous_level]][empty_values], "NA", sep="_")
    tx_table[[level]][empty_values] <- gsub("(_NA[> ]*)*", "\\1", tx_table[[level]][empty_values]) # Replace repetitions of NA by one NA

    previous_level <- level
  }

  return(tx_table)
}
taxmat <- as.matrix(keep_unidentified(taxmat))
colnames(taxmat) <- stringr::str_replace(colnames(taxmat), "^\\w{1}", toupper)

create.table(head(taxmat), "Head of taxonomy table.")
```

# **Create phyloseq object**

The phyloseq object can finally be created merging the feature, taxonomy and metadata tables:

```{r message=FALSE, warning=FALSE}
OTU = otu_table(otu_df, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)
SAM = sample_data(metadata)

physeq = phyloseq(OTU, TAX, SAM)

physeq = subset_samples(physeq, origin=="faeces" & sample.type!="control" & exclusion!="excluded")
sample_data(physeq)[c("region_16s","sample.type.label","database.label")] <- (data.frame(sample_data(physeq)) %>% dplyr::mutate(
  region_16s = case_when(
    platform == "Illumina" ~ "Illumina-V3V4",
    platform == "Oxford_Nanopore" ~ "ONT-V1V9",
    TRUE ~ ""
  ),
  sample.type.label = case_when(
    sample.type == "crc" ~ "Cancer",
    sample.type == "non-crc" ~ "Control",
    TRUE ~ ""
  ),
  database.label = case_when(
    database == "silva" ~ "SILVA", # SILVA
    database == "default" ~ "Default", # "rrnDB+NCBI" "EmuDefault"
    TRUE ~ ""
  )
))[c("region_16s","sample.type.label","database.label")]

physeq

```

Additionally, we remove possible contaminant features using their genus and known contaminants from the input "contamination_file", with some additional manual names specified in the code:

```{r}
pop_taxa = function(ps, bad_asvs){
  all_asvs = taxa_names(ps)
  keep_these <- all_asvs[!(all_asvs %in% bad_asvs)]
  return(prune_taxa(keep_these, ps))
}
# Only bacteria
bad_asvs <- taxa_names(subset_taxa(tax_table(physeq), Domain!="Bacteria"))
physeq <- pop_taxa(physeq, bad_asvs)

# Remove typical contamination
typical_contamination <- read.csv(contamination_file, sep='\t')

cont_genus <- unique(c(
  # Base genus
  typical_contamination$Genus,
  paste("g__", typical_contamination$Genus, sep=""),
  # Genus with NA
  paste(paste("g__", typical_contamination$Genus, sep=""), "NA", sep= "_"),
  paste(paste("g__", typical_contamination$Genus, sep=""), "NA", sep= " "),
  paste(typical_contamination$Genus, "NA", sep= "_"),
  paste(typical_contamination$Genus, "NA", sep= " ")
))
physeq <- subset_taxa(physeq, !Genus %in% cont_genus)

# Manual names
temp <- c("Dermacoccaceae",
          "Mitochondria",
          "Chloroplast")
if (any(temp %in% data.frame(tax_table(physeq))$Family)){
  bad_asvs <- taxa_names(
    subset_taxa(
      tax_table(physeq),
      Family %in% temp #|
        # Genus %in% c(
        #   "g__Burkholderia-Caballeronia-Paraburkholderia",
        #   "g__Methylobacterium-Methylorubrum",
        #   "g__Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium",
        #   "g__Acinetobacter"
        # )
    )
  )
  physeq <- pop_taxa(physeq, bad_asvs)
}

# Keep ASVs with more than one count
physeq <- prune_taxa(taxa_sums(physeq) > 1, physeq)
physeq
```

# **Calculate rarefied phyloseq object**

A rarefied phyloseq object is calculated here and will be used in some posterior analysis sucha as beta-diversity:

```{r}
rar_physeq <- rarefy_even_depth(physeq, sample.size = 15000,
                                replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
rar_physeq
```

<!-- # **Data export** -->

<!-- Export physeq object: -->

<!-- ```{r} -->
<!-- save.physeq <- function(physeq_obj, output_file, add_taxonomy=TRUE){ -->
<!--   # Keep taxa with at least one count (not all zeroes) -->
<!--   physeq_obj <- prune_taxa(taxa_sums(physeq_obj) >= 1, physeq_obj) -->

<!--   # Merge taxonomy with ASVs -->
<!--   tax_cols <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species") -->
<!--   merged.taxa.otus <- merge(otu_table(physeq_obj), tax_table(physeq_obj)[,tax_cols], by.x = 0, by.y = 0, all.x = TRUE, all.y = TRUE) -->
<!--   names(merged.taxa.otus)[1] <- "feature_id" -->

<!--   # Merge tax into one column -->
<!--   merged.taxa.otus$taxonomy <- do.call(paste, c(merged.taxa.otus[tax_cols], sep=";")) -->
<!--   for (c in tax_cols) {merged.taxa.otus[c] <- NULL} -->

<!--   # Make feature_id into row index -->
<!--   rownames(merged.taxa.otus) <- merged.taxa.otus$feature_id -->

<!--   # Count total -->
<!--   # merged.taxa.otus$TOTAL <- rowSums(merged.taxa.otus[,-which(colnames(merged.taxa.otus) %in% c("taxonomy","feature_id"))]) -->

<!--   # Reorder columns -->
<!--   merged.taxa.otus <- merged.taxa.otus[,c("feature_id","taxonomy", -->
<!--                                           colnames(merged.taxa.otus)[!colnames(merged.taxa.otus) %in% c("taxonomy","feature_id")] -->
<!--   ) -->
<!--   ] -->
<!--   # merged.taxa.otus <- merged.taxa.otus[order(-merged.taxa.otus$TOTAL),] -->
<!--   if (add_taxonomy==FALSE){ -->
<!--     merged.taxa.otus$taxonomy <- NULL -->
<!--   } -->

<!--   # Save -->
<!--   write.table(merged.taxa.otus, file=output_file, sep='\t', row.names=F) -->
<!-- } -->
<!-- temp <- subset_samples(physeq, platform=="Oxford_Nanopore" & basecalling_model_name == "sup") -->
<!-- temp <- prune_taxa(taxa_sums(temp) >= 1, temp) -->
<!-- save.physeq(temp, glue("{output_folder}/clean_feature_table.tsv"), add_taxonomy=TRUE) -->
<!-- write.table(data.frame(sample_data(temp)), glue("{output_folder}/clean_feature_table_metadata.tsv"), sep="\t", row.names=F) -->

<!-- ``` -->

# **Alpha-diversity**

Using the rarefied phyloseq object we obtain alpha-diversity measures. First, the main function is described:

```{r}
plot_alpha_diversity <- function(physeq_obj, measures, x, color_col){
  # ---- Plot alpha diversity ----
  alpha_div_plot <- plot_richness(physeq_obj, x=x, color=color_col, measures=measures) +
    geom_boxplot() + theme_bw() +
    # facet_grid(variable~origin, scale="free") +
    # scale_x_discrete(limits = c("saliva__non-crc", "saliva__crc", "subgingival-fluid__crc", "faeces__non-crc", "faeces__crc", "normal-mucosa__crc", "adenocarcinoma__crc")) +
    theme(plot.title = element_text(hjust = 0.5),
          panel.grid.minor.y = element_blank(),
          axis.title = element_text(size = 12),
          axis.text.x = element_text(angle = 90, hjust=0.95, vjust=0.2),
          axis.text = element_text(size = 12))

  # ---- Wilcoxon-test ----
  richness <- estimate_richness(physeq_obj, measures=measures)
  rownames(richness) <- sample_data(physeq_obj)$sample.id
  # rownames(richness) <- gsub("\\.", '-', rownames(richness))
  richness <- merge(richness, data.frame(sample_data(physeq_obj)), by=0)
  richness$group <- as.factor(richness$group)

  # Run test for each combination of groups
  wilcox_list <- list()
  n = 1
  for (m in measures){
    for (i in sort(unique(richness$group))){
      for (j in sort(unique(richness$group))){
        res <- NULL
        # As long as the group is not the same (error if it is)
        if (i!=j){
          # Run the test between groups and print if significative
          res <- wilcox.test(as.formula(glue("{m} ~ group")),
                             data=subset(richness,
                                         group %in% c(i, j)
                             )
          )
          wilcox_list[[n]] <- c(i, j, m, res$p.value)
        } else {
          wilcox_list[[n]] <- c(i, j, m, NA)
        }
        n=n+1
      }
    }
  }

  # Format Wilcox results
  wilcox_df <- as.data.frame(do.call(rbind, wilcox_list))
  names(wilcox_df) <- c("C1","C2","measure","pval")
  wilcox_df$pval <- as.numeric(wilcox_df$pval)

  # Plot each wilcoxon test
  significance_p_list <- list()
  for (m in measures){
    df <- subset(wilcox_df, measure==m)[c("C1","C2","pval")]

    # Turn to wide and block lower triangle
    df <- as.data.frame(df %>% tidyr::pivot_wider(names_from = C1, values_from = pval))
    df[-1][lower.tri(df[-1])] <- NA

    # Turn to long
    df <- df %>% pivot_longer(!C2, names_to = "C1", values_to = "pval")

    p <- ggplot(df, aes(C1, C2, fill=pval)) +
      geom_tile() +
      scale_fill_gradient(low="darkgreen", high="white", na.value = "white", limits=c(0,0.1)) +
      # scale_y_discrete(position = "right") +
      theme_bw() +
      ylab("") + ggtitle(glue("{m}")) + xlab("") +
      theme(axis.text.x=element_text(angle = 60, hjust=0.95, vjust=0.95), legend.key.size = unit(0.5,"line"))
    significance_p_list[[m]] <- p
  }

  return(list(alpha_div_plot=alpha_div_plot, significance_p_list=significance_p_list, richness=richness[,c("sample.id",measures)]))

}

```

Calculate alpha-diversity at the species level for all samples:

```{r}
measures <- c("ACE","Shannon","Simpson","Observed","Chao1")
ps <- tax_glom(subset_samples(rar_physeq), "Species")
sample_data(ps)$group <- (data.frame(sample_data(ps)) %>% unite("temp", all_of(variables_of_interest), sep=" - "))$temp

res = plot_alpha_diversity(ps, measures, "group", "sample.type")

temp <- res$richness %>%
  tidyr::pivot_longer(!sample.id, names_to = "alphadiversity_measure", values_to = "alphadiversity_value")
temp <- merge(temp, data.frame(sample_data(ps)), on="sample.id")
```

Plot ONT's alpha-diversity with both databases:

```{r}
model_alphadiversity_p <- ggplot(
    subset(temp, alphadiversity_measure %in% c("Observed","Shannon","Simpson") & platform=="Oxford_Nanopore") %>%
      dplyr::mutate(x_axis=glue("{sample.type.label}\n({database.label})")),
    aes(x=x_axis, alphadiversity_value, fill=basecalling_model_name)
  ) +
  geom_boxplot(alpha=0.5, outlier.shape = NA) +
  geom_point(aes(color=basecalling_model_name), position=position_jitterdodge(jitter.width=0.2), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
  facet_wrap(~alphadiversity_measure, scales="free") +
  theme_bw() +
  xlab(NULL) +
  ylab("Alpha-diversity") +
  # ggtitle("") +
  guides(fill=guide_legend(title="Basecalling model")) +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(angle = 0, hjust=0.5, vjust=0.5, size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold")
  ) +
  ggpubr::geom_pwc(
    aes(group = basecalling_model_name), step.increase=0.05, vjust = 0.5, tip.length = 0.01,
    method = "wilcox_test", label = "{p.adj.signif}",
    p.adjust.method = "holm", hide.ns = TRUE
  ) #+
  # ggpubr::geom_pwc(
  #   aes(group = x_axis), step.increase=0.05, vjust = 0.5, size = 0.75, tip.length = 0,
  #   method = "wilcox_test", label = "{p.adj.signif}",
  #   p.adjust.method = "holm", hide.ns = TRUE,
  #   bracket.nudge.y = 0.2
  # )

model_alphadiversity_p
ggsave(glue("{output_folder}/alphadiversity_ont.png"), plot=model_alphadiversity_p, dpi="retina", unit="px", height=2000, width=3000)

```

Plot ONT Sup vs. Illumina with SILVA:

```{r}
platform_alphadiversity_p <- ggplot(
    subset(temp, alphadiversity_measure %in% c("Observed","Shannon","Simpson") & database == "silva" & (basecalling_model_name == "sup" | platform=="Illumina")),
    aes(x=sample.type.label, y=alphadiversity_value, fill=region_16s)
  ) +
  geom_boxplot(alpha=0.5, outlier.shape = NA) +
  geom_point(position=position_jitterdodge(jitter.width=0.1), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
  facet_wrap(~alphadiversity_measure, scales="free", nrow=1) +
  ggpubr::geom_pwc(
    aes(group = region_16s), step.increase=0.05, vjust = 0.5, tip.length = 0.01,
    method = "wilcox_test", label = "{p.adj.signif}",
    p.adjust.method = "holm", hide.ns = TRUE
  ) +
  theme_bw() +
  xlab(NULL) +
  ylab("Alpha-diversity") +
  # ggtitle("") +
  guides(fill=guide_legend(title="Method   ")) +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(angle = 0, hjust=0.5, vjust=0.5, size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold")
  )
platform_alphadiversity_p
ggsave(glue("{output_folder}/alphadiversity_platforms.png"), plot=platform_alphadiversity_p, dpi="retina", unit="px", height=2000, width=2500)

```

# **Beta-diversity**

Beta-diveristy comparisons for ONT basecalling models (silva + default) at the species level:

```{r}
rar_ps = tax_glom(subset_samples(rar_physeq, platform == "Oxford_Nanopore"), "Species")
dist <- ordinate(rar_ps, "MDS", "jsd", formula=as.formula(glue("~ {paste(variables_of_interest, collapse = ' + ')}")))
sample_data(rar_ps)$temp <- paste(sample_data(rar_ps)$biosample_name, sample_data(rar_ps)$database.label, sep="_")
sample_data(rar_ps)$model_database <- paste(sample_data(rar_ps)$basecalling_model_name, sample_data(rar_ps)$database.label, sep=" - ")

basecalling_model_silvaanddefault_jsd <- plot_ordination(rar_ps, dist,
                                         type="sites", color="model_database") +
  geom_path(aes(group=temp), color="black", alpha=0.7, linetype=1, linewidth=0.5) +
  guides(fill=guide_legend(title="Basecalling model")) +
  theme_bw() +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(size=10),
    plot.title=element_text(face="bold", size=17)
  )
basecalling_model_silvaanddefault_jsd
```

Beta-diveristy comparisons for ONT basecalling models (silva) at the species level:

```{r}
rar_ps = tax_glom(subset_samples(rar_physeq, database == "silva" & platform == "Oxford_Nanopore"), "Species")
dist <- ordinate(rar_ps, "MDS", "jsd", formula=as.formula(glue("~ {paste(variables_of_interest, collapse = ' + ')}")))

basecalling_model_silva_jsd <- plot_ordination(rar_ps, dist,
                                         type="sites", color="basecalling_model_name") +
  geom_path(aes(group=biosample_name), color="black", alpha=0.7, linetype=1, linewidth=0.5) +
  guides(fill=guide_legend(title="Basecalling model")) +
  theme_bw() +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(size=10),
    plot.title=element_text(face="bold", size=17)
  )
basecalling_model_silva_jsd
```

Beta-diversity comparison for Illumina vs Nanopore at the genus level:

```{r}
rar_ps = tax_glom(subset_samples(rar_physeq, database == "silva" & (basecalling_model_name == "sup" | platform == "Illumina")), "Genus")
dist <- ordinate(rar_ps, "MDS", "jsd", formula=as.formula(glue("~ {paste(variables_of_interest, collapse = ' + ')}")))

platform_genus_silva_jsd <- plot_ordination(rar_ps, dist,
                                         type="sites", color="region_16s") +
  stat_ellipse() +
  guides(fill=guide_legend(title="Method")) +
  theme_bw() +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(size=10),
    plot.title=element_text(face="bold", size=17)
  )
platform_genus_silva_jsd
```

Beta-diversity comparison for Illumina vs Nanopore at the species level:

```{r}
rar_ps = tax_glom(subset_samples(rar_physeq, database == "silva" & (basecalling_model_name == "sup" | platform == "Illumina")), "Species")
dist <- ordinate(rar_ps, "MDS", "jsd", formula=as.formula(glue("~ {paste(variables_of_interest, collapse = ' + ')}")))

platform_species_silva_jsd <- plot_ordination(rar_ps, dist,
                                      type="sites", color="region_16s") +
  stat_ellipse() +
  guides(fill=guide_legend(title="Method")) +
  theme_bw() +
  theme(
    axis.text=element_text(size=10, color="black"),
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.text=element_text(size=10),
    legend.title=element_text(size=10, face="bold"),
    axis.text.x=element_text(size=10),
    plot.title=element_text(face="bold", size=17)
  )
platform_species_silva_jsd
```

# **PERMANOVA using adonis2**

PERMANOVA at the genus level using JSD:

```{r}
# Filter
rar_ps = tax_glom(subset_samples(rar_physeq), "Genus")

# Matriz de distancias
rar_physeq_dist <- phyloseq::distance(rar_ps, "jsd")
sampledf <- data.frame(sample_data(rar_ps))

# Adonis test
permanova <- adonis2(
  as.formula(paste("rar_physeq_dist ~ ", paste(variables_of_interest, collapse="+"), sep = "")), # rar_physeq_dist ~ origin + sample.type,
  data = sampledf, permutations = 9999, by="onedf",
  parallel = 60)

temp <- data.frame(permanova)
write.table(data.frame(variable = row.names(temp), temp), glue("{output_folder}/permanova.tsv"), sep="\t", row.names=F)
print(temp)
```

# **CLR Abundance**

Define functions:

```{r}
plot_boxes <- function(data, x, y, facet="target_level", fill=NULL, title=NULL, boxplot.outlier.shape = 19){
  ggplot(data, aes_string(x=x, y=y, fill=fill)) +
    facet_wrap(as.formula(glue("~{facet}")), scales="free") +
    geom_boxplot(alpha=0.5, outlier.shape=boxplot.outlier.shape) +
    # geom_dotplot(binaxis='y', stackdir='center', dotsize=.5, alpha=0.5) +
    theme_bw() +
    xlab(NULL) +
    ylab("Abundance (CLR)") +
    ggtitle(title) +
    guides(fill=guide_legend(title="")) +
    theme(
      axis.text=element_text(size=10, color="black"),
      axis.title=element_text(size=10, face="bold"),
      axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
      legend.text=element_text(size=10),
      axis.text.x=element_text(angle = 0, hjust=0.5, vjust=0.5, size=10),
      plot.title = element_text(size = 15, color = 'black', face="bold"),
      strip.background = element_rect(fill="white", color="black"),
      strip.text = element_text(colour = 'black', face="bold"),
      strip.text.x = element_text(colour = 'black', face="bold.italic"),
    )
}

like.with.vector.filter <- function(data, searches, column_to_search){
  data <- data[
    rowSums(# Get count of times each pattern appears
      sapply( # Search pattern in column
        sapply(searches, function(x){glue(".*{x}.*")}), # create vector of taxa like ".*Fusobacterium.*" #ORAL_GENERA,
        like, vector = data[[column_to_search]])
    )>=1, # If a pattern appears, select row
  ]
  return(data)
}

get.clr.abund.at.zero <- function(rel_abund_data, clr_abund_data){
  # Merge CLR and Relative abundance long objects and get the
  # point at which clr abundance is mostly 0% relative abundance

  # Rename columns
  rel_abund_data <- rel_abund_data %>% dplyr::rename(relative_abundance=Abundance)
  clr_abund_data <- clr_abund_data %>% dplyr::rename(clr_abundance=Abundance)

  # Merge
  merge_df <- merge(
    rel_abund_data[c("sample.id","grouping_var","OTU","target_level","relative_abundance")],
    clr_abund_data[c("sample.id","grouping_var","OTU","target_level","clr_abundance")],
    by=c("sample.id","grouping_var","OTU","target_level")
  )

  # Format
  merge_df$relative_abundance <- merge_df$relative_abundance*100
  merge_df$relative_is_zero <- merge_df$relative_abundance==0

  # Plot correspondence
  p <- ggplot(data=merge_df, aes_string(x="relative_abundance", y="clr_abundance", color="relative_is_zero")) +
    geom_point() +
    ggtitle("CLR abundance vs Relative abundance") +
    geom_hline(yintercept = max(subset(merge_df, relative_abundance==0)$clr_abundance), alpha=0.2) +
    theme_bw()
  # ggplotly(p)

  # Get point at which clr abundance is mostly 0% relative abundance
  clr.abund.at.zero <- max(subset(merge_df, relative_abundance==0)$clr_abundance)

  return(clr.abund.at.zero)
}

get.long.clr.and.rel <- function(physeq_obj, selected_taxa_level){
  # Normalize
  physeq_obj <- tax_glom(physeq_obj, selected_taxa_level)
  physeq_obj <- prune_taxa(taxa_sums(physeq_obj) >= 1, physeq_obj)

  clr_ps <- microbiome::transform(physeq_obj, transform="clr")
  rel_abund_ps <- microbiome::transform(physeq_obj, transform="compositional")

  # Agglomerate and turn into long
  clr_ps_long <- clr_ps %>%
    tax_glom(taxrank = selected_taxa_level, bad_empty = c()) %>%   # Aglomerate to taxnomic range
    psmelt()
  clr_ps_long$target_level <- clr_ps_long[[selected_taxa_level]]

  rel_abund_ps_long <- rel_abund_ps %>%
    tax_glom(taxrank = selected_taxa_level, bad_empty = c()) %>%   # Aglomerate to taxnomic range
    psmelt()
  rel_abund_ps_long$target_level <- rel_abund_ps_long[[selected_taxa_level]]

  # Format
  clr_ps_long[["target_level"]] <- gsub(".*__", '', clr_ps_long[["target_level"]])
  clr_ps_long[["target_level"]] <- gsub("_", ' ', clr_ps_long[["target_level"]])
  rel_abund_ps_long[["target_level"]] <- gsub(".*__", '', rel_abund_ps_long[["target_level"]])
  rel_abund_ps_long[["target_level"]] <- gsub("_", ' ', rel_abund_ps_long[["target_level"]])

  return(list(clr=clr_ps_long, rel_abund=rel_abund_ps_long))
}
```

Process each combination of CLR normalization:

```{r}
# ---- CLR at certain level ----
# For model sup, database silva at Species level
l <- get.long.clr.and.rel(
  subset_samples(physeq, database == "silva" & (basecalling_model_name=="sup" )), #| platform=="Illumina"
  selected_taxa_level <- "Species"
)
clr_ps_long_sp <- l$clr
rel_abund_ps_long_sp <- l$rel_abund

# For ONT sup with default database at species level
l <- get.long.clr.and.rel(
  subset_samples(physeq, database == "default" & (basecalling_model_name=="sup")),
  selected_taxa_level <- "Species"
)
clr_ps_long_default_sp <- l$clr
rel_abund_ps_long_default_sp <- l$rel_abund

# For all models from ONT, database silva at Species level
l <- get.long.clr.and.rel(
  subset_samples(physeq, database == "silva" & platform=="Oxford_Nanopore"),
  selected_taxa_level <- "Species"
)
clr_ps_long_ont_models_sp <- l$clr
rel_abund_ps_long_ont_models_sp <- l$rel_abund

# For all models from ONT, database silva at genus level
l <- get.long.clr.and.rel(
  subset_samples(physeq, database == "silva" & platform=="Oxford_Nanopore"),
  selected_taxa_level <- "Genus"
)
clr_ps_long_ont_models_genus <- l$clr
rel_abund_ps_long_ont_models_genus <- l$rel_abund

# For ONT vs Illumina, database silva at the genus level
l <- get.long.clr.and.rel(
  subset_samples(physeq, database == "silva" & (basecalling_model_name=="sup"| platform=="Illumina")),
  selected_taxa_level <- "Genus"
)
clr_ps_long_genus <- l$clr
rel_abund_ps_long_genus <- l$rel_abund

# For Illumina (silva) and ONT sup (with both databases) at the genus level
# both databases are used at the same time because we only want relative abundance, not CLR
l <- get.long.clr.and.rel(
  subset_samples(physeq, database %in% c("silva","default") & (basecalling_model_name=="sup"| platform=="Illumina")),
  selected_taxa_level <- "Genus"
)
rel_abund_ps_long_databases_genus <- l$rel_abund

# For comparing both platforms and both databases with CLR
temp <- list(
  get.long.clr.and.rel(
    subset_samples(physeq, database == "silva" & basecalling_model_name=="sup"),
    selected_taxa_level <- "Genus"
  ),
  get.long.clr.and.rel(
    subset_samples(physeq, database == "default" & basecalling_model_name=="sup"),
    selected_taxa_level <- "Genus"
  ),
  get.long.clr.and.rel(
    subset_samples(physeq, database == "silva" & platform=="Illumina"),
    selected_taxa_level <- "Genus"
  )
)
clr_ps_long_platform <- dplyr::bind_rows(temp[[1]]$clr, temp[[2]]$clr, temp[[3]]$clr)
rel_abund_ps_long_platform <- dplyr::bind_rows(temp[[1]]$rel_abund, temp[[2]]$rel_abund, temp[[3]]$rel_abund)

```

Plot specific genera, Illumina vs ONT. Note: Three different CLR objects are calculated separatedly, as the CLR normalization seems to be influenced by the amount of feature_ids with values equal to 0, and because databases were different the taxonomy IDs were also different.

```{r fig.width=15}
clr.abund.at.zero <- -0.4

clr_ps_long_platform <- clr_ps_long_platform %>%
  dplyr::mutate(
    temp_col = factor(
      glue("{region_16s}\n{sample.type.label}\n{database.label}"),
      levels=c(
        "Illumina-V3V4\nCancer\nSILVA", "ONT-V1V9\nCancer\nSILVA", "ONT-V1V9\nCancer\nDefault",
        "Illumina-V3V4\nControl\nSILVA", "ONT-V1V9\nControl\nSILVA", "ONT-V1V9\nControl\nDefault"
      )
    )
  )
temp <- subset(clr_ps_long_platform, target_level %in% c("Parvimonas", "Fusobacterium", "Peptostreptococcus"))
clr_genus_platform <- plot_boxes(
  temp,
  x="temp_col", y="Abundance", facet="target_level", fill="sample.type.label",
  boxplot.outlier.shape = NA # outliers drawn by next geom_point jitter
) +
  # geom_path(aes(group=biosample_name), color="black", alpha=0.6, linetype=2, linewidth=0.2) +
  geom_point(position=position_jitterdodge(jitter.width=0.5), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
  geom_hline(yintercept=clr.abund.at.zero, color="black", alpha=0.5) +
  annotate(
    "rect", xmin = -Inf, xmax = Inf, ymin = -Inf,
    ymax = clr.abund.at.zero,
    fill = "dodgerblue4", alpha = 0.1, color = NA
  ) +
  geom_signif(
    comparisons = list(
      c("ONT-V1V9\nCancer\nDefault", "ONT-V1V9\nControl\nDefault"),
      c("ONT-V1V9\nCancer\nSILVA", "ONT-V1V9\nControl\nSILVA"),
      c("Illumina-V3V4\nCancer\nSILVA",  "Illumina-V3V4\nControl\nSILVA")
    ),
    map_signif_level=map_signif_level_func,
    step_increase=0.05,
    textsize = 3,
    tip_length = 0.01,
    margin_top = 0.1,
    vjust = 0.5
  )
clr_genus_platform
```

Plot specific species, only ONT with both databases. Note: Two different objects, one with database=silva and another with database=default are calculated separatedly, as the CLR normalization seems to be influenced by the amount of feature_ids with values equal to 0, and because databases were different the taxonomy IDs were also different.

```{r fig.width=15}

clr.abund.at.zero <- mean(
  get.clr.abund.at.zero(rel_abund_ps_long_sp, clr_ps_long_sp),
  get.clr.abund.at.zero(rel_abund_ps_long_default_sp, clr_ps_long_default_sp)
)

temp <- subset(
  dplyr::bind_rows(clr_ps_long_sp, clr_ps_long_default_sp) %>%
    dplyr::mutate(x_label = glue("{sample.type.label}\n{database.label}")) %>%
    dplyr::mutate(x_label = factor(x_label, levels=c("Cancer\nDefault", "Control\nDefault", "Cancer\nSILVA","Control\nSILVA")))  %>%
    dplyr::mutate(
      target_level = dplyr::case_when(
        target_level == "Clostridium sensu stricto 1 perfringens" ~ "Clostridium perfringens",
        target_level == "Clostridium sensu stricto 1 saudiense" ~ "Clostridium saudiense",
        TRUE ~ target_level
      )
    ),
  target_level %in% c(
    # Species
    "Parvimonas micra",
    "Fusobacterium nucleatum", 
    # "Fusobacterium gonidiaformans", "Fusobacterium necrophorum",
    "Bacteroides fragilis",
    "Peptostreptococcus stomatis",
    "Peptostreptococcus anaerobius", "Sutterella wadsworthensis",
    "Paraprevotella clara", "Dialister pneumosintes",
    "Clostridium perfringens",
    "Gemella morbillorum",
    "Clostridium saudiense",
    # "Ruficoccus amylovorans",
    # Probable good species
    "Agathobaculum butyriciproducens",
    "Romboutsia ilealis",
    "Anaerostipes rhamnosivorans",
    "Anaerocolumna cellulosilytica"
  )
)
clr_sp_ont_both_p <- plot_boxes(
  temp,
  x="x_label", y="Abundance", facet="target_level", fill="x_label", boxplot.outlier.shape=NA # outliers are drawn by geom_point with jitter
) +
  facet_wrap(~target_level, scales="free_y", nrow=3) +
  geom_point(position=position_jitterdodge(jitter.width=1), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
  geom_hline(yintercept=clr.abund.at.zero, color="black", alpha=0.5) +
  scale_fill_brewer(palette = "Dark2") +
  annotate(
    "rect", xmin = -Inf, xmax = Inf, ymin = -Inf,
    ymax = clr.abund.at.zero,
    fill = "dodgerblue4", alpha = 0.1, color = NA
  ) +
  # ggpubr::geom_pwc(
  #   aes(group = x_label), data=temp%>%dplyr::filter(database.label=="Default"),
  #   step.increase=0.05, vjust = 0.5, tip.length = 0.01,
  #   method = "wilcox_test", label = "{p.adj.signif}",
  #   p.adjust.method = "holm", hide.ns = TRUE
  # ) +
  # ggpubr::geom_pwc(
  #   aes(group = x_label), data=temp%>%dplyr::filter(x_label %in% c("Cancer\nSILVA","Control\nSILVA")),
  #   step.increase=0.05, vjust = 0.5, tip.length = 0.01,
  #   method = "wilcox_test", label = "{p.adj.signif}",
  #   p.adjust.method = "holm", hide.ns = TRUE
  # )
  geom_signif(
    comparisons = list(
      c("Cancer\nDefault","Control\nDefault"),
      c("Cancer\nSILVA","Control\nSILVA")
    ),
    map_signif_level=map_signif_level_func,#map_signif_level_func,
    step_increase=0, #0.05,
    textsize = 3,
    tip_length = 0.01,
    margin_top = 0.1,
    vjust = 0.5
  )
clr_sp_ont_both_p

```

<!-- Plot all species with significant adjusted p-value when comparing subjects in Default database (ONTV1V9 sup): -->

<!-- ```{r} -->

<!-- # Calculate p-values -->

<!-- wilcoxon.res <- clr_ps_long_default_sp %>%  -->

<!--   dplyr::group_by(platform, basecalling_model_name, database, target_level) %>% -->

<!--   rstatix::wilcox_test(Abundance ~ sample.type.label) %>% -->

<!--   rstatix::adjust_pvalue(method = "holm") %>% -->

<!--   dplyr::filter(p <= 0.1) -->

<!-- clr_ps_long_subset <- like.with.vector.filter(clr_ps_long_default_sp, c(unique(wilcoxon.res$target_level)), "target_level") -->

<!-- rel_abund_ps_long_subset <- like.with.vector.filter(rel_abund_ps_long_default_sp, c(unique(wilcoxon.res$target_level)), "target_level") -->

<!-- # ---- Plot into n split columns ---- -->

<!-- # Select taxa -->

<!-- target.taxa <- sort(unique(clr_ps_long_subset$target_level)) -->

<!-- # Check if any of them dont exist in data -->

<!-- target.taxa[!target.taxa %in% clr_ps_long_subset$target_level] -->

<!-- # Get point under which CLR abundance is mostly equivalent to 0% relative abundance -->

<!-- clr.abund.at.zero <- get.clr.abund.at.zero(rel_abund_ps_long_subset, clr_ps_long_subset) -->

<!-- # Plot -->

<!-- targets.groups <- split(target.taxa, ceiling(seq_along(target.taxa)/40)) -->

<!-- p_list <- list() -->

<!-- n = 1 -->

<!-- for (g in targets.groups){ -->

<!--   p <- plot_boxes( -->

<!--       subset( -->

<!--         clr_ps_long_subset,target_level %in% g), -->

<!--       x="sample.type.label", y="Abundance", facet="Species", fill="sample.type.label", boxplot.outlier.shape = NA # outliers are placed by next geom_point jitter -->

<!--     ) + -->

<!--     geom_point(position=position_jitterdodge(jitter.width=0.5), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) + -->

<!--     # facet_grid(target_level~platform, scales="free") + -->

<!--     # geom_vline(xintercept = c(4.5, 8.5, 12.5), alpha=0.6, linetype="dotted") + -->

<!--     geom_hline(yintercept=clr.abund.at.zero, color="black", alpha=0.5) + -->

<!--     annotate( -->

<!--       "rect", xmin = -Inf, xmax = Inf, ymin = -Inf, -->

<!--       ymax = clr.abund.at.zero, -->

<!--       fill = "dodgerblue4", alpha = 0.1, color = NA -->

<!--     ) + geom_signif( -->

<!--       comparisons = list( -->

<!--         c("Control","Cancer") -->

<!--       ), -->

<!--       map_signif_level=map_signif_level_func, -->

<!--       step_increase=0.05, -->

<!--       textsize = 3, -->

<!--       tip_length = 0.01, -->

<!--       margin_top = 0.1, -->

<!--       vjust = 0.5 -->

<!--     ) -->

<!--   # p -->

<!--   ggsave( -->

<!--     glue("{output_folder}/all_default_p{n}.png"),  -->

<!--     plot=p, dpi="retina", units="px", height=4000, width=6000, limitsize = FALSE -->

<!--   ) -->

<!--   p_list[[n]] <- p -->

<!--   n = n+1 -->

<!-- } -->

<!-- ``` -->

# **Differential abundance**

Define functions:

```{r}
library(ANCOMBC)
run_ancombc_1vs1 <- function(physeq_obj, target_level, target_variables, prv_cut=0.3, qval_min=0.1){
  # IMPORTANT: ANCOM-BC v2.0.1 was used. Its output format varies depending on version
  if (target_level=="ASV"){
    ancom_target <- NULL
  } else {
    ancom_target <- target_level
  }

  # Run ANCOM-BC
  out <- ancombc(data=physeq_obj, tax_level=ancom_target, prv_cut=prv_cut,
                 formula = paste(target_variables, collapse="+"), n_cl=60,
                 p_adj_method="holm", alpha = 0.05)
  # Get results
  res = out$res

  # To long
  lfc_long <- res$lfc[,c(-2)] %>% tidyr::pivot_longer(cols = starts_with(target_variables), names_to = "var", values_to = "lfc")
  diff_abn_long <- res$diff_abn[,c(-2)] %>% tidyr::pivot_longer(cols = starts_with(target_variables), names_to = "var", values_to = "diff_abn")
  se_long <- res$se[,c(-2)] %>% tidyr::pivot_longer(cols = starts_with(target_variables), names_to = "var", values_to = "se")
  qval_long <- res$q_val[,c(-2)] %>% tidyr::pivot_longer(cols = starts_with(target_variables), names_to = "var", values_to = "qval")

  # Merge long
  merged <- Reduce(function(x, y) merge(x, y, by=c("taxon","var")),
                   list(lfc_long, diff_abn_long, se_long, qval_long))

  # Get taxons where any of the tests achived a certain qval
  selected_taxons <- res$q_val[rowSums(res$q_val[,c(-1,-2),drop=F]<=qval_min)!=0,]$taxon
  merged <- subset(merged, taxon %in% selected_taxons)
  merged

  # Change column name
  names(merged)[1] <- "taxon_id"

  # Add names if ASV was selected as target_level
  if (target_level=="ASV"){
    merged$taxon_id <- data.frame(tax_table(physeq_obj))[merged$taxon_id,][,"Species"]
  }

  # Format taxon_id to remove things like "g__"
  # merged$taxon_id <- gsub(".*__", '', merged$taxon_id)

  # Format results for plotting
  df_fig <- merged %>%
    dplyr::arrange(desc(lfc)) %>%
    dplyr::mutate(direction = ifelse(lfc > 0, "Positive LFC", "Negative LFC"))
  df_fig$direction <- factor(df_fig$direction, levels = c("Positive LFC", "Negative LFC"))

  # Significance label
  # - Corrected p-value (qval) under 0.001 will have "***"
  # - Corrected p-value (qval) under 0.01 will have "**"
  # - Corrected p-value (qval) under 0.05 will have "*"
  df_fig$qval.txt <- cut(df_fig$qval, c(-Inf,0.001,0.01,0.05,Inf), labels=c('***','**','*','-'))

  # Orientation for significance level
  df_fig$orientation <- 0
  df_fig$orientation[df_fig$lfc > 0] <- 1
  df_fig$orientation[df_fig$lfc < 0] <- -1

  # Add target level
  df_fig$target_level <- target_level

  return(df_fig)
}

# Function to get the proper alpha value for each significance level
# (non-significant entries will be grayed out)
get_scale_alpha_values <- function(vec){
  l <- length(unique(vec))
  alpha_vec <- rep(1, l)
  if ("-" %in% vec){
    alpha_vec[length(alpha_vec)] <- 0.4
  }
  return(alpha_vec)
}

plot.ancombc <- function(data, fill="direction"){
  p <- ggplot(
    data = subset(
      data %>%
        dplyr::arrange(lfc) %>%
        dplyr::mutate(taxon_id=factor(taxon_id, levels=unique(.$taxon_id)))
      # dplyr::filter(grepl('Parvimonas|Fusobacterium|Peptostreptococcus', taxon_id))
    ),
    aes_string(x = "taxon_id", y = "lfc", fill=fill, alpha="qval.txt")
  ) +
    scale_alpha_manual(values=get_scale_alpha_values(data$qval.txt)) +
    geom_bar(stat = "identity", color="black", linewidth=0.3,
             position = position_dodge2(width = 0.9, preserve = "single")) +
    geom_errorbar(aes(ymin = lfc - se, ymax = lfc + se), width=0.5,
                  position = position_dodge2(width = 0.9, preserve = "single"),
                  color = "gray36") +
    geom_text(aes(label = qval.txt, y=lfc+orientation*(se+0.22)), vjust = 0.7, color = "gray36",
              position = position_dodge2(width = 0.9, preserve = "single")) +
    # geom_vline(xintercept = (1:length(data$taxon_id)-1)+0.5, alpha=0.5, linetype="dotted") +
    geom_hline(yintercept = 0, alpha=0.5) +
    facet_wrap(~target_level, scale="free_y", ncol=1) +
    labs(x = NULL, y = "Log fold change", title = "") +
    coord_flip() +
    guides(alpha = "none") +
    theme_bw() +
    theme(
      axis.text=element_text(size=10, color="black"),
      axis.text.x=element_text(angle = 0, hjust=0.95, vjust=0.95, size=10),
      axis.text.y=element_text(size=10, face="italic"),
      axis.title=element_text(size=10, face = "bold"),
      legend.text=element_text(size=8),
      # panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill="white", color="black"),
      strip.text = element_text(colour = 'black', face="bold"),
      strip.text.y = element_text(colour = 'black', face="bold.italic")
    )
  return(p)
}
```

Define prevalence cut:

```{r}
ancombc_prv_cut <- 0.1
```

ONTV1V9 (sup) Default database, cancer vs control subjects:

```{r}
# Filter phyloseq object
derep_ps <- subset_samples(physeq, sample.type %in% c("crc","non-crc") & database=="default" & (basecalling_model_name == "sup"))
derep_ps <- prune_taxa(taxa_sums(derep_ps) >= 1, derep_ps)
sample_data(derep_ps)$sample.type.label <- factor(sample_data(derep_ps)$sample.type.label, levels=c("Control", "Cancer"))

# Run on each level
proc_genus_ancombc <- NULL
proc_species_ancombc <- NULL

# proc_genus_ancombc <- run_ancombc_1vs1(derep_ps, "Genus", c("sample.type.label"), prv_cut=ancombc_prv_cut)
proc_species_ancombc <- run_ancombc_1vs1(derep_ps, "Species", c("sample.type.label"), prv_cut=ancombc_prv_cut)

# Concatenate results
ont.default.df_fig <- rbind(proc_genus_ancombc, proc_species_ancombc)
ont.default.df_fig$target_level <- factor(ont.default.df_fig$target_level,      # Reordering group factor levels
                              levels = c("Phylum","Class","Order","Family","Genus","Species","ASV")
)
create.table(ont.default.df_fig)
ancombc.ont.default.p <- plot.ancombc(
      subset(ont.default.df_fig, target_level=="Species") %>% 
        dplyr::mutate(  
          direction = case_when(
            direction == "Positive LFC" ~ "More in Cancer  ",
            direction == "Negative LFC" ~ "More in Control  ",
            TRUE ~ ""
          )
        )
    ) + 
      theme(
        strip.background = element_blank(), 
        strip.text.x = element_blank(), 
        legend.text=element_text(size=12), 
        strip.text=element_text(size=10),
        plot.margin = margin(0,0.25,0,2, "cm")
    ) +
      guides(fill=guide_legend(title=NULL))
ancombc.ont.default.p
```

ONTV1V9 (sup) with SILVA database, cancer vs control subjects:

```{r}
# Filter phyloseq object
derep_ps <- subset_samples(physeq, sample.type %in% c("crc","non-crc") & database=="silva" & (basecalling_model_name == "sup"))
derep_ps <- prune_taxa(taxa_sums(derep_ps) >= 1, derep_ps)
sample_data(derep_ps)$sample.type.label <- factor(sample_data(derep_ps)$sample.type.label, levels=c("Control", "Cancer"))

# Run on each level
proc_genus_ancombc <- NULL
proc_species_ancombc <- NULL

# proc_genus_ancombc <- run_ancombc_1vs1(derep_ps, "Genus", c("sample.type.label"), prv_cut=ancombc_prv_cut)
proc_species_ancombc <- run_ancombc_1vs1(derep_ps, "Species", c("sample.type.label"), prv_cut=ancombc_prv_cut)

# Concatenate results
ont.silva.df_fig <- rbind(proc_genus_ancombc, proc_species_ancombc)
ont.silva.df_fig$target_level <- factor(ont.silva.df_fig$target_level,      # Reordering group factor levels
                              levels = c("Phylum","Class","Order","Family","Genus","Species","ASV")
)
create.table(ont.silva.df_fig)
ancombc.ont.silva.p <- plot.ancombc(
      subset(ont.silva.df_fig, target_level=="Species") %>% 
        dplyr::mutate(  
          direction = case_when(
            direction == "Positive LFC" ~ "More in Cancer  ",
            direction == "Negative LFC" ~ "More in Control  ",
            TRUE ~ ""
          ),
          taxon_id = gsub(" NA$", " sp.", taxon_id)
        )
    ) + 
      theme(
        strip.background = element_blank(), 
        strip.text.x = element_blank(), 
        legend.text=element_text(size=12), 
        strip.text=element_text(size=10),
        plot.margin = margin(0,0.25,0,2, "cm")
    ) +
      guides(fill=guide_legend(title=NULL))
ancombc.ont.silva.p
```

# **Random forest with Boruta and manual feature selection**

Set up:

```{r}
boruta_folder <- "/home/usuario/Proyectos/Results/ONT16S/pipeline/16_boruta"
```

Automatic Boruta feature selection with each database and two prevalence cuts:

```{r}
prev_cuts <- c(0.1,0.3)
dbs <- c("default","silva")


auc_l <- list()
boruta_l <- list()
for (d in dbs){
  for (p in prev_cuts){
    boruta_f <- glue("{boruta_folder}/results.boruta.boruta.species.{p}.{d}.tsv")
    auc_f <- glue("{boruta_folder}/results.RF.boruta.species.{p}.{d}.tsv")
    
    boruta_df <- data.table::fread(boruta_f) %>% 
      dplyr::mutate(db=d, prev_cut=p)
    auc_df <- data.table::fread(auc_f) %>% 
      dplyr::rename(feature_count=n.vars, auc=AUC) %>%
      dplyr::mutate(db=d, prev_cut=as.character(p*100)) %>%
      dplyr::mutate(
          db = case_when(
            db == "default" ~ "Default",
            db == "silva" ~ "SILVA",
            TRUE ~ ""
          )
      ) %>%
      dplyr::group_by(feature_count)%>%
      dplyr::mutate(taxa=paste(boruta_df$feature[0:feature_count], collapse="\n"))
      
    boruta_l[[glue("{d}_{p}")]] <- boruta_df
    auc_l[[glue("{d}_{p}")]] <- auc_df
  }
}

auc_df <- do.call(rbind,auc_l)
boruta_df <- do.call(rbind,boruta_l)


automatic_boruta_p <- ggplot(auc_df, aes(x=feature_count, y=auc, color=glue("{db} ({prev_cut}%)"))) + #, linetype=db
  geom_line(show.legend = FALSE) +
  geom_point(alpha=0.5, size=2) +
  # facet_wrap(~db)+
  # geom_area(alpha=0.2) +
  # geom_label_repel(
  #   aes(label=taxa, fill=NULL),
  #   subset(auc_df, auc == 0.9 & feature_count <= 10),
  #   color="black", segment.colour="gray50", show.legend = FALSE, 
  #   nudge_x=-13, nudge_y=-0.05, hjust=0
  # ) +
  scale_x_reverse(breaks = scales::pretty_breaks(n = 10)) +
  theme_bw() +
  xlab("Features") +
  ylab("AUC") +
  ggtitle(NULL) +
  guides(color=guide_legend(title="Model")) +
  theme(
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    # axis.text.x=element_text(angle = 0, hjust=1, vjust=0, size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.y = element_text(colour = 'black', face="bold.italic"),
    # panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
automatic_boruta_p
```

Manual feature selection (ANCOM-BC + Boruta + CLR abundance + identity analysis):

```{r}
# Process data
auc_comp <- data.table::fread(glue("{boruta_folder}/auc_illu_vs_ont.tsv")) %>%
  dplyr::mutate(old_method_features=gsub(", ", "\n", old_method_features),
                new_method_features=gsub(", ", "\n", new_method_features),
                combination=row_number()) %>%
  tidyr::pivot_longer(cols = contains("_method_"),
               names_pattern = "(\\w+)_method_(\\w+)",
               names_to = c("method", ".value")) %>%
  dplyr::mutate(feature_count=str_count(features,"\n")+1) %>%
  dplyr::mutate(
    method = case_when(
      method == "old" ~ "Illumina-V3V4",
      method == "new" ~ "ONT-V1V9",
      TRUE ~ ""
    ),
    auc_label = glue("**{auc}**<br>(n={feature_count})")
  )

# Create annotations
annotation <- subset(auc_comp, str_count(groups, ":")==1)$groups
annotation <- sub("(G\\d+:)(.+) ", "**\\1**\\2", annotation)
annotation <- paste(
  annotation,
  collapse="<br>"
)


manual_auc_p <- ggplot(auc_comp, aes(x=combination, y=auc, color=method, fill=method)) +
  geom_line(linetype=2) +
  geom_point(aes(size=feature_count), show.legend = FALSE) +
  geom_richtext(
    aes(label=auc_label),
    data=subset(auc_comp, method=="Illumina-V3V4"),
    hjust=0.5, vjust=1.22,
    color="black",
    fill = NA, label.color = NA
  ) +
  geom_richtext(
    aes(label=auc_label),
    data=subset(auc_comp, method=="ONT-V1V9"),
    hjust=0.5, vjust=-0.28,
    color="black",
    fill = NA, label.color = NA
  ) +
  geom_richtext(
    aes(label=glue('**{gsub(":.*", "", groups)}**')),
    data=subset(auc_comp, str_count(groups, ":")==1),
    hjust=0.5, vjust=-2.5,
    color="black",
    fill = NA, #label.color = NA
  ) +
  geom_textbox(
    aes(x = 16, y = 0.78), label = annotation, 
    color="black", fill="white", box.colour="gray60",
    width = grid::unit(0.45, "npc"), # 73% of plot panel width
    hjust = 0, vjust = 1, show.legend = FALSE
  ) +
  scale_x_reverse(breaks = scales::pretty_breaks(n = 10)) +
  ylim(0.68,0.91) + 
  # geom_label(aes(label=auc, size=feature_count), color="white", show.legend = FALSE) +
  theme_bw() +
  xlab("Combination") +
  ylab("AUC") +
  ggtitle(NULL) +
  guides(color=guide_legend(title="Model")) +
  theme(
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    # axis.text.x=element_text(angle = 0, hjust=1, vjust=0, size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.y = element_text(colour = 'black', face="bold.italic"),
    # panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
manual_auc_p
```

# **Proportion of feature counts classified perfectly as a described species**

```{r PerfectSpeciesProportion, cache.lazy = FALSE}
selected_taxa_level = "Species"
temp <- physeq %>%
  tax_glom(taxrank = selected_taxa_level, bad_empty = c()) %>%   # Aglomerate to taxnomic range
  psmelt()
temp$target_level <- temp[[selected_taxa_level]]


pattern <- " NA$|uncultured|unidentified|unclassified|human gut|^bacterium | sp.$|gut metagenome|metagenome"
temp <- temp %>%
  # Establish taxa not matching pattern
  dplyr::mutate(
    taxa_is_identified = !grepl(pattern, target_level)
  ) %>%
    # Group by and sum counts according to if taxa is identified
    dplyr::group_by(biosample_name, platform, region_16s, basecalling_model_name, database, database.label, sample.type.label, taxa_is_identified) %>%
    dplyr::summarise(sum_abundance=sum(Abundance)) %>%
    pivot_wider(names_from = taxa_is_identified, values_from = sum_abundance, names_prefix="taxa_is_identified_") %>%
    dplyr::mutate(
      identified_pct = 100 * taxa_is_identified_TRUE/(taxa_is_identified_FALSE+taxa_is_identified_TRUE),
      unidentified_pct = 100 * taxa_is_identified_FALSE/(taxa_is_identified_FALSE+taxa_is_identified_TRUE)
    ) %>%
    pivot_longer(cols=c(identified_pct, unidentified_pct), names_to = "type", values_to = "pct")

temp <- temp %>% dplyr::filter(type=="identified_pct") %>% dplyr::mutate(label=glue("{region_16s}\n{database.label}\n{basecalling_model_name}"))
temp$label <- factor(
  temp$label, 
  levels=c(
    "Illumina-V3V4\nSILVA\n",  "ONT-V1V9\nSILVA\nfast", "ONT-V1V9\nSILVA\nhac", "ONT-V1V9\nSILVA\nsup",
    "ONT-V1V9\nDefault\nfast","ONT-V1V9\nDefault\nhac", "ONT-V1V9\nDefault\nsup"
    )
)

temp_gg_fun <- function(data){
  p <- ggplot(
      data,
      aes(x=label, y=pct, fill=label)
    ) + 
    geom_boxplot(alpha=0.7, outlier.shape=NA) +
    geom_point(position=position_jitterdodge(jitter.width=0.25), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
    facet_wrap(~sample.type.label) + 
    ggpubr::geom_pwc(
      aes(group = label), step.increase=0.05, vjust = 0.5, tip.length = 0.01,
      method = "wilcox_test", label = "{p.adj.signif}",
      p.adjust.method = "holm", hide.ns = TRUE
    ) +
    ylab(glue("Feature counts percentage properly identified at described {selected_taxa_level}")) +
    xlab(NULL) +
    guides(fill=guide_legend(title=NULL)) +
    # scale_y_continuous(expand = c(0,0), limits=c(0,110)) +
    theme_bw() +
    theme(legend.position="null") +
    theme(
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text=element_text(size=10, color="black"), 
      axis.title=element_text(size=10, face="bold"),
      axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
      legend.title=element_text(size=10, face="bold"),
      legend.text=element_text(size=10),
      plot.title = element_text(size = 15, color = 'black', face="bold"),
      strip.background = element_rect(fill="white", color="black"), # #3263b0
      strip.text = element_text(colour = 'black', face="bold"),
      panel.background = element_blank(),
      axis.ticks.x=element_blank()
    )
  return(p)
}
species_level_idenfitication_percentages_silva_p <- temp_gg_fun(temp %>% dplyr::filter(database=="silva"))
species_level_idenfitication_percentages_silva_p
ggsave(glue("{output_folder}/supplementary_fig_2.png"), plot=species_level_idenfitication_percentages_silva_p, dpi="retina", units="px", width=3000, height=2200)

species_level_idenfitication_percentages_p <- temp_gg_fun(temp)
species_level_idenfitication_percentages_p

create.table(temp %>% 
  dplyr::group_by(label) %>%
  rstatix::get_summary_stats() %>%
  dplyr::filter(variable=="pct") %>%
  dplyr::select(label, n, min, max, median, mean, sd, se), "Stats on percentage properly identified as described Species")
```


# **Emu reads classification comparison**

In the following plots the accuracy of each basecalling model in respect to the best one (sup) is measured, showing the percentage of reads identified differently:

```{r}

get.matched.pct.per.sample <- function(filepath, selected_taxa_level){
  data <- read.delim(filepath) %>%
    dplyr::rename(level_0=glue("{selected_taxa_level}_0"), level_1=glue("{selected_taxa_level}_1")) %>%
    dplyr::group_by(comparison_id, level_0, level_1) %>%
    dplyr::summarise(count=sum(read_id)) %>%
    dplyr::mutate(
      matches=case_when(
        level_0 == level_1 ~ "match",
        TRUE ~ "nomatch"
      )
    ) %>%
    dplyr::group_by(comparison_id, matches) %>%
    dplyr::summarise(count=sum(count)) %>%
    tidyr::pivot_wider(names_from="matches", values_from="count", values_fill=0) %>%
    dplyr::summarise(non_matching_reads_pct=100*nomatch/(match+nomatch)) %>%
    dplyr::mutate(level=selected_taxa_level) 
  
  return(data)
}


temp_folder = "/home/usuario/Proyectos/Results/ONT16S/pipeline/10.2_compare_emu_reads"
data <- dplyr::bind_rows(
  # hac vs sup silva
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_silva.tsv"), "family") %>% dplyr::mutate(comp="hac vs sup (SILVA)", level="Family"),
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_silva.tsv"), "genus") %>% dplyr::mutate(comp="hac vs sup (SILVA)", level="Genus"),
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_silva.tsv"), "species") %>% dplyr::mutate(comp="hac vs sup (SILVA)", level="Species"),
  # fast vs sup silva
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_silva.tsv"), "family") %>% dplyr::mutate(comp="fast vs sup (SILVA)", level="Family"),
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_silva.tsv"), "genus") %>% dplyr::mutate(comp="fast vs sup (SILVA)", level="Genus"),
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_silva.tsv"), "species") %>% dplyr::mutate(comp="fast vs sup (SILVA)", level="Species"),
  # hac vs sup default
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_default.tsv"), "family") %>% dplyr::mutate(comp="hac vs sup (Default)", level="Family"),
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_default.tsv"), "genus") %>% dplyr::mutate(comp="hac vs sup (Default)", level="Genus"),
  get.matched.pct.per.sample(glue("{temp_folder}/hac_sup_default.tsv"), "species") %>% dplyr::mutate(comp="hac vs sup (Default)", level="Species"),
  # fast vs sup default
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_default.tsv"), "family") %>% dplyr::mutate(comp="fast vs sup (Default)", level="Family"),
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_default.tsv"), "genus") %>% dplyr::mutate(comp="fast vs sup (Default)", level="Genus"),
  get.matched.pct.per.sample(glue("{temp_folder}/fast_sup_default.tsv"), "species") %>% dplyr::mutate(comp="fast vs sup (Default)", level="Species")
)


# stat.test <- data %>% 
#   dplyr::group_by(comp) %>%
#   rstatix::wilcox_test(non_matching_reads_pct ~ level) %>%
#   rstatix::adjust_pvalue(method = "holm") %>%
#   # rstatix::add_xy_position(x = "comp", group = "level") %>%
#   dplyr::filter(p.adj<=0.1) 

emu_reads_comparison_p <- ggplot(data, aes(x=level, y=non_matching_reads_pct, fill=comp)) +
  geom_boxplot(alpha=0.5, outlier.shape=NA) +
  geom_point(position=position_jitterdodge(jitter.width=0.1), color="black", shape = 21, stroke = 0.5, show.legend=FALSE, size=1, alpha=0.7) +
  ggpubr::geom_pwc(
    aes(group = comp), step.increase=0.05, vjust = 0.5, tip.length = 0.01,
    method = "wilcox_test", label = "{p.adj.signif}",
    p.adjust.method = "holm", hide.ns = TRUE
  ) +
  # ggpubr::geom_pwc(
  #   aes(group = comp), step.increase=0.05, vjust = 0.5, tip.length = 0,
  #   method = "wilcox_test", label = "{p.adj.signif}",
  #   p.adjust.method = "holm", hide.ns = TRUE,
  #   bracket.nudge.y = 0.2, size=0.75
  # ) +
  scale_fill_brewer(palette="Dark2") +
  ylab("Percentage of reads identified differently") +
  xlab(NULL) +
  guides(fill=guide_legend(title="Comparison     ")) +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"), # #3263b0
    strip.text = element_text(colour = 'black', face="bold"),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
emu_reads_comparison_p
ggsave(glue("{output_folder}/supplementary_fig_1.png"), plot=emu_reads_comparison_p, dpi="retina", units="px", width=3000, height=2200)

create.table(
  data %>%
    dplyr::group_by(comp, level) %>%
    rstatix::get_summary_stats() %>%
    dplyr::filter(variable=="non_matching_reads_pct") %>%
    dplyr::select(level, comp, n, min, max, median, mean, sd, se),
  "Boxplot stats on percentage of reads identified differently"
)

```

# **Paper plots**

## Figure 1

```{r}
# ---- Read quality stats plot ----
# Read
sup <- data.table::fread(file="/home/usuario/Proyectos/Results/ONT16S/pipeline/09_qualities/sup.tsv.gz")
hac <- data.table::fread(file="/home/usuario/Proyectos/Results/ONT16S/pipeline/09_qualities/hac.tsv.gz")
fast <- data.table::fread(file="/home/usuario/Proyectos/Results/ONT16S/pipeline/09_qualities/fast.tsv.gz")

# Rename
names(sup) <- c("length","avg_quality")
names(hac) <- c("length","avg_quality")
names(fast) <- c("length","avg_quality")

# Filter
sup <- sup %>% dplyr::filter(avg_quality>=12 & length>=1200 & length<=1900)
hac <- hac %>% dplyr::filter(avg_quality>=12 & length>=1200 & length<=1900)
fast <- fast %>% dplyr::filter(avg_quality>=7 & length>=1200 & length<=1900)

# Add basecalling model
sup$basecalling_model_name <- "sup"
hac$basecalling_model_name <- "hac"
fast$basecalling_model_name <- "fast"

# Merge
temp <- rbind(sup, hac, fast)
temp.stats <- temp %>% 
  dplyr::group_by(basecalling_model_name) %>% 
  dplyr::summarise(
    median_length=median(length), 
    median_avg_quality=round(median(avg_quality)), 
    mean_avg_quality=round(mean(avg_quality)), 
    reads=n(),
    reads_m=scales::label_number(accuracy=0.1, scale_cut=scales::cut_short_scale())(n())
  )

# Sample n rows to reduce size of table
temp <- dplyr::sample_n(temp, dim(temp)[1]/10, replace = TRUE)

# Create annotations
annotation <-apply(temp.stats, 1, function(x){glue("- {x['basecalling_model_name']}: Q{x['median_avg_quality']}, {x['median_length']}bp, {x['reads_m']}\n")})
annotation <- paste(
  annotation,
  collapse="\n"
)

# Plot
read_quality_stats_p <- ggplot(temp, aes(x=avg_quality, fill=basecalling_model_name)) +
  # geom_histogram(aes(y=..density..), bins = 100, alpha=0.5, position="identity") +
  geom_density(alpha=0.5) +
  annotate("label", x = 25, y = 0.35, hjust=0, size=4,
           label.padding=unit(0.7, "lines"), label.r = unit(0.4, "lines"),
           label = annotation
  ) +
  theme_bw() +
  xlab("Quality") +
  ylab("Density") +
  ggtitle(NULL) +
  guides(fill=guide_legend(title="Basecalling model")) +
  scale_y_continuous(expand = c(0,0)) + 
  scale_x_continuous(expand = c(0,0)) + 
  theme(
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    # axis.text.x=element_text(angle = 0, hjust=1, vjust=0, size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.y = element_text(colour = 'black', face="bold.italic"),
    # panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
read_quality_stats_p <- read_quality_stats_p + geom_magnify(
  from = c(xmin = 25, xmax = max(temp$avg_quality), ymin = 0, ymax = 0.005), 
  to = c(xmin = 27, xmax = 37, ymin = 0.05, ymax = 0.25),
  corners = 0.1, axes = "xy"
)
# ggsave(glue("{output_folder}/reads_quality.png"), plot=read_quality_stats_p, dpi="retina", units="px", width=3500, height=2500)
rm(temp, sup, hac, fast) # remove large objects

# -- Plot --
p <- ggpubr::ggarrange(
  ggpubr::ggarrange(
    read_quality_stats_p + theme(legend.position="none"), 
    basecalling_model_silva_jsd + theme(legend.position="none"), 
    widths=c(0.6,0.4),
    ncol = 2, labels = c("A", "B")#, common.legend=TRUE, legend="bottom"
  ),
  ggpubr::ggarrange(
    model_alphadiversity_p + theme(legend.position="bottom") + guides(fill=guide_legend(title="Basecalling model     ")),
    # emu_reads_comparison_p + theme(legend.position="bottom"),
    # clr_model_correspondence_p,
    ncol=1,
    labels = c("C")#, common.legend=TRUE, legend="bottom"
  ),
  nrow=2, #heights=c(0.25,0.75),
  font.label = list(size = 30), common.legend=TRUE, legend="bottom"
) + ggpubr::bgcolor("white") + ggpubr::border("white")
p
ggsave(glue("{output_folder}/possible_fig_1.png"), plot=p, dpi="retina", units="px", width=4200, height=3500)
```

## Figure 2

```{r}
# -- CLR Correspondence between platforms --
temp_cols <- c("Domain", "Phylum", 
               "Class", "Order", "Family", "Genus","sample.type.label")

x <- clr_ps_long_genus %>% 
  dplyr::group_by(platform, basecalling_model_name, across(temp_cols)) %>% 
  dplyr::summarise(mean_abundance=mean(Abundance))
clr.abund.at.zero <- get.clr.abund.at.zero(rel_abund_ps_long_genus, clr_ps_long_genus)

x <- merge(
  subset(x[c(temp_cols,"platform","basecalling_model_name","mean_abundance")], platform=="Oxford_Nanopore" & basecalling_model_name=="sup"),
  subset(x[c(temp_cols,"platform","basecalling_model_name","mean_abundance")], platform=="Illumina"),
  by=c(temp_cols)
) %>% 
  dplyr::rename(mean_abundance_ont=mean_abundance.x, mean_abundance_illumina=mean_abundance.y)

x <- x[c("sample.type.label","Family","Genus","mean_abundance_ont","mean_abundance_illumina")]
x$concat_taxa <- paste(x$Family, x$Genus, sep=";")

thresh <- clr.abund.at.zero
x <- x %>% dplyr::mutate(
  appears = case_when(
    mean_abundance_ont > thresh & mean_abundance_illumina > thresh ~ "Both",
    mean_abundance_ont > thresh & mean_abundance_illumina <= thresh ~ "Only ONT",
    mean_abundance_ont <= thresh & mean_abundance_illumina > thresh ~ "Only Illumina",
    mean_abundance_ont <= thresh & mean_abundance_illumina <= thresh ~ "None",
    TRUE ~ ""
  )
)
x <- subset(x, Family!="uncultured" & !Family %like% "Unassigned")

# Check unequal taxonomy
x_subset <- subset(x, (appears %like% "Only" | appears=="None") & sample.type.label=="Cancer")
for (i in unique(x_subset$Genus)){
  if (!endsWith(i, " NA") & dim(subset(x_subset, Genus %like% glue("{i} NA")))[1]>=1){
    # print("##################")
    # print(i)
    # x <- subset(x, !Genus %like% i) ### WARNING
    # print(subset(x_subset, Genus %like% i))
  }
}

# x <- subset(x, sample.type.label=="Cancer")
clr_technology_correspondence_p <- ggplot(x, aes(x=mean_abundance_ont, y=mean_abundance_illumina, color=appears, fill=appears)) +
  geom_point(alpha=0.5, size = 2, color="black", shape = 21, stroke = 0.5) + 
  geom_smooth(aes(fill=appears), data=subset(x, appears=="Both"), method = "lm", se = TRUE, alpha=0.08, linetype=3, show.legend=FALSE) +
  geom_label_repel(
    aes(label=Genus),
    data=subset(x, sample.type.label=="Control" & Genus %in% c("Fusobacterium","Parvimonas","Peptostreptococcus")),
    segment.colour="gray50", show.legend = FALSE, nudge_x=2, nudge_y=6.5, fontface = "bold.italic", color="black", size = 3
  ) +
  geom_label_repel(
    aes(label=Genus),
    data=subset(x, sample.type.label=="Cancer" & Genus %in% c("Fusobacterium","Parvimonas","Peptostreptococcus")),
    segment.colour="gray50", show.legend = FALSE, nudge_x=1, nudge_y=5, fontface = "bold.italic", color="black", size = 3
  ) +
  stat_cor(
    aes(label =  paste(..rr.label.., ..p.label.., sep = "~`,`~")), # paste(..rr.label.., ..p.label.., sep = "~`,`~")
    data=subset(x, appears=="Both"),
    method = "pearson", cor.coef.name="Pearson: R", p.accuracy = 0.001, r.accuracy = 0.01,
    label.x=3, label.y=7.5, show.legend = FALSE
  ) +
  facet_wrap(~sample.type.label) +
  guides(fill=guide_legend(title="Taxon appears\non average on")) +
  xlab("CLR Abundance (ONT)") +
  ylab("CLR Abundance (Illumina)") +
  theme_bw() +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  theme(
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.y = element_text(colour = 'black', face="bold"),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
# clr_technology_correspondence_p



p <- ggpubr::ggarrange(
  ggpubr::ggarrange(
    platform_alphadiversity_p + 
      theme(legend.position="none"),
      # scale_fill_brewer(palette="Dark2"),
    platform_genus_silva_jsd + 
      # geom_path(aes(group=biosample_name), color="black", alpha=0.6, linetype=2, linewidth=0.1) + 
      theme(legend.position="none"),
      # scale_color_brewer(palette="Dark2"), 
    platform_species_silva_jsd + 
      # geom_path(aes(group=biosample_name), color="black", alpha=0.6, linetype=2, linewidth=0.1) + 
      theme(legend.position="none"),
      # scale_color_brewer(palette="Dark2"), 
    ncol = 3, labels = c("A", "B", "C"), widths=c(0.45,0.27,0.27), common.legend=TRUE, legend="right"#, hjust = -0.2
  ),
  ggpubr::ggarrange(
    clr_technology_correspondence_p,
    labels = c("D"), legend="right"#, hjust = -0.05
  ),
  nrow=2,
  font.label = list(size = 30)
) + ggpubr::bgcolor("white") + ggpubr::border("white")   
ggsave(glue("{output_folder}/possible_fig_2.png"), plot=p, dpi="retina", units="px", width=4200, height=2800)
p

```

## Figure 3

```{r}
temp <- rel_abund_ps_long_databases_genus[c("sample.id","biosample_name","sample.type.label","platform","database.label","region_16s","Genus","Abundance")] %>%
  dplyr::filter(Genus %in% c("Parvimonas", "Fusobacterium", "Peptostreptococcus")) %>%
  dplyr::filter(sample.id %in% sample_data(rar_physeq)$sample.id) %>%
  dplyr::group_by(biosample_name, sample.type.label, region_16s, database.label, Genus) %>%
  dplyr::summarise(sum_abundance=sum(100*Abundance), sample_count=n()) %>%
  tidyr::pivot_wider(names_from = c(region_16s, database.label), values_from = sum_abundance) %>% 
  dplyr::arrange(biosample_name, Genus) %>%
  dplyr::filter(!is.na(`ONT-V1V9_SILVA`) & !is.na(`Illumina-V3V4_SILVA`) & !is.na(`ONT-V1V9_Default`)) # if one of the values is NA that sample is not available for both technologies, remove

# -- Percentage of samples with taxon present per group --
pct.per.class <- temp %>%
  dplyr::group_by(Genus, sample.type.label) %>%
  dplyr::summarise(
    "Illumina-V3V4 SILVA" = round(100 * sum(`Illumina-V3V4_SILVA`!=0) / n(), 2), 
    "ONT-V1V9 SILVA" = round(100 * sum(`ONT-V1V9_SILVA`!=0) / n(), 2),
    "ONT-V1V9 Default" = round(100 * sum(`ONT-V1V9_Default`!=0) / n(), 2),
  )

# pct.per.class.table.p <- ggtexttable(
#     pct.per.class %>% 
#       dplyr::rename("Sample type" = sample.type.label), 
#     rows = NULL, theme = ttheme("blank")
#   ) %>%
#   tab_add_hline(at.row = 1:2, row.side = "top", linewidth = 2)



pct.per.class.p <- ggplot(
  pct.per.class %>%   
    tidyr::pivot_longer(
      cols=c("Illumina-V3V4 SILVA","ONT-V1V9 SILVA","ONT-V1V9 Default"), 
      names_to = "platform_database", values_to = "presence_pct"
    ) %>% 
    dplyr::mutate(platform_database=gsub(" ","\n",platform_database)), 
  aes(x=platform_database, y=presence_pct, fill=sample.type.label)
) + 
  geom_bar(stat="identity", position=position_dodge(0.9), alpha=0.6, width = 0.8, color="black") +
  geom_text(aes(label = paste(presence_pct, "%", sep="")), hjust = -0.1, size=3, position = position_dodge(0.9)) +
  facet_wrap(~Genus, ncol=1, strip.position="right") +
  ylab("% of samples with taxon") +
  xlab(NULL) +
  guides(fill=guide_legend(title=NULL)) +
  scale_y_continuous(expand = c(0,0), limits=c(0,55)) +
  coord_flip() +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.y = element_text(colour = 'black', face="bold.italic"),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
# pct.per.class.p

# -- Relative abundance per taxon and sample --
temp <- temp %>% 
  dplyr::mutate(non_zero_presence_count=sum(c(`Illumina-V3V4_SILVA`!=0, `ONT-V1V9_Default`!=0, `ONT-V1V9_SILVA`!=0)), na.rm = TRUE) %>%
  tidyr::pivot_longer(cols=c("Illumina-V3V4_SILVA","ONT-V1V9_Default","ONT-V1V9_SILVA"), names_to = "platform_database", values_to = "abundance") %>%
  dplyr::arrange(desc(abundance)) %>% 
  dplyr::mutate(biosample_name=factor(biosample_name, levels=unique(.$biosample_name)))  %>%
  dplyr::na_if(0)
# temp

sample.matrix.p <- ggplot(
    temp %>% dplyr::mutate(platform_database=gsub("_","\n",platform_database)), 
    aes(x=platform_database, y=biosample_name, fill=abundance)
  ) + 
  geom_tile(show.legend = FALSE) +
  scale_fill_distiller(palette = "YlOrRd", na.value="white", direction=1) +
  geom_text(aes(label = round(abundance, 4)), size=2.5) +
  facet_grid(sample.type.label~Genus, scales="free", space = "free") +
  xlab(NULL) +
  ylab("Samples") +
  scale_x_discrete(expand = c(0,0)) +
  theme_bw() +
  theme(
    # axis.text.y=element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text=element_text(size=10, color="black"), 
    axis.title=element_text(size=10, face="bold"),
    axis.title.y=element_text(size=10, face="bold", margin = margin(t = 0, r = 5, b = 0, l = 0)),
    legend.title=element_text(size=10, face="bold"),
    legend.text=element_text(size=10),
    plot.title = element_text(size = 15, color = 'black', face="bold"),
    strip.background = element_rect(fill="white", color="black"),
    strip.text = element_text(colour = 'black', face="bold"),
    strip.text.x = element_text(colour = 'black', face="bold.italic"),
    panel.background = element_blank(),
    axis.ticks.x=element_blank()
  )
# sample.matrix.p


# -- Figure 3 --
p <- ggpubr::ggarrange(
  clr_genus_platform + facet_wrap(~target_level, scales="free_y", nrow=3,  strip.position="right") + theme(legend.text = element_text(size=12)),
  pct.per.class.p + theme(legend.position="bottom") + theme(legend.text = element_text(size=12)),
  ncol=2,
  labels=c("A","B"),
  font.label = list(size = 20),
  common.legend=TRUE,legend="bottom"
) + ggpubr::bgcolor("white") + ggpubr::border("white")
ggsave(glue("{output_folder}/possible_fig_3.png"), plot=p, dpi="retina", units="px", width=4500, height=2700)
p


```

## Figure 4

```{r fig.width=12, fig.height=13}
p <- ggpubr::ggarrange(
  ggpubr::ggarrange(
    ancombc.ont.silva.p,
    ancombc.ont.default.p,
    labels=c("A","B"),
    font.label = list(size = 20),
    common.legend=TRUE, legend="bottom"
  ),
  ggpubr::ggarrange(
    clr_sp_ont_both_p + 
      theme(legend.text=element_text(size=12), 
            strip.text=element_text(size=9.5)),
    font.label = list(size = 20),
    labels=c("C"), legend="bottom"
  ),
  nrow=2, heights=c(0.4,0.6)
) + ggpubr::bgcolor("white") + ggpubr::border("white")
ggsave(glue("{output_folder}/possible_fig_4.png"), plot=p, dpi="retina", units="px", width=4800, height=4500)
p

```


## Figure 4v2

```{r fig.width=12, fig.height=13}
p <- ggpubr::ggarrange(
  ggpubr::ggarrange(
    ancombc.ont.silva.p,
    ancombc.ont.default.p,
    labels=c("A","B"),
    font.label = list(size = 20),
    common.legend=TRUE, legend="bottom"
  ),
  ggpubr::ggarrange(
    clr_sp_ont_both_p + 
      theme(legend.text=element_text(size=12), 
            strip.text=element_text(size=9.5)),
    manual_auc_p + 
      theme(
        legend.spacing.x=unit(0.3, "cm"), 
        legend.text = element_text(size=12, hjust = 0.5)
      ) + guides(color=guide_legend(title=NULL)),
    font.label = list(size = 20),
    labels=c("C","D"), legend="bottom",
    nrow=2
  ),
  nrow=2, heights=c(0.3,0.7)
) + ggpubr::bgcolor("white") + ggpubr::border("white")
ggsave(glue("{output_folder}/possible_fig_4v2.png"), plot=p, dpi="retina", units="px", width=4800, height=6500)
p

```

## Figure 5

```{r fig.height  = 10}
p <- ggpubr::ggarrange(
  automatic_boruta_p + 
    theme(
      legend.spacing.x=unit(0.3, "cm"), 
      legend.text = element_text(size=12, hjust = 0.5),
    ) + guides(color=guide_legend(title=NULL))
  ,
  manual_auc_p + 
    theme(
      legend.spacing.x=unit(0.3, "cm"), 
      legend.text = element_text(size=12, hjust = 0.5)
    ) + guides(color=guide_legend(title=NULL))
  ,
  nrow = 2, labels = c("A", "B"), common.legend=FALSE, legend="bottom",
  font.label = list(size = 15)
) + ggpubr::bgcolor("white") + ggpubr::border("white")   
ggsave(glue("{output_folder}/possible_fig_5.png"), plot=p, dpi="retina", units="px", width=4200, height=4000)
p
```

# **Session info**

Session info, including used packages, is available here:

```{r}
sessionInfo()
```
